cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+class ExtractedPathImpl : public ExtractedPath {
+ public:
+  std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>> paths;
+  ExtractedPathImpl() {
+    if (std::is_same<T, double>::value) {
+      cond_type = CondType::kDouble;
+    } else {
+      cond_type = CondType::kFloat;
+    }
+  }
+  virtual ~ExtractedPathImpl() = default;
+
+  CondType GetCondType() override {
+    return cond_type;
+  }
+};
+
+template <typename T>
+std::unique_ptr<ExtractedPath> ExtractedPath::Create() {
+  std::unique_ptr<ExtractedPath> model
+    = std::make_unique<ExtractedPathImpl<T>>();
+  return model;
+}
+
+template <typename Func, typename ...Args>
+auto
+ExtractedPath::Dispatch(Func func, Args&& ...args) {
+  switch (this->cond_type) {
+   case CondType::kDouble:
+    func(*dynamic_cast<ExtractedPathImpl<double>*>(this),
+         std::forward<Args>(args)...);
+    break;
+   case CondType::kFloat:
+   default:
+    func(*dynamic_cast<ExtractedPathImpl<float>*>(this),
+         std::forward<Args>(args)...);
+    break;
+  }
+}
+
+template <typename T>
+std::ostream& operator<<(
+    std::ostream& os,
+    const std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>>& paths) {
+  std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>> tmp(paths);
+  std::sort(tmp.begin(), tmp.end(),
+            [&](const gpu_treeshap::PathElement<MySplitCondition<T>>& a,
+                const gpu_treeshap::PathElement<MySplitCondition<T>>& b) {
+              if (a.path_idx < b.path_idx) return true;
+              if (b.path_idx < a.path_idx) return false;
+
+              if (a.feature_idx < b.feature_idx) return true;
+              if (b.feature_idx < a.feature_idx) return false;
+              return false;
+            });
+
+  for (auto i = 0ull; i < tmp.size(); i++) {
+    auto e = tmp[i];
+    if (i == 0 || e.path_idx != tmp[i - 1].path_idx) {
+      os << "path_idx:" << e.path_idx << ", leaf value:" << e.v;
+      os << "\n";
+    }
+    os << " (feature:" << e.feature_idx << ", pz:" << e.zero_fraction << ", ["
+       << e.split_condition.feature_lower_bound << "<=x<"
+       << e.split_condition.feature_upper_bound << "])";
+    os << "\n";
+  }
+  return os;
+}
+
+class DenseDatasetWrapper {
+  const float* data;
+  std::size_t num_rows;
+  std::size_t num_cols;
+
+ public:
+  DenseDatasetWrapper() = default;
+  DenseDatasetWrapper(const float* data, int num_rows, int num_cols)
+      : data(data), num_rows(num_rows), num_cols(num_cols) {}
+  __device__ float GetElement(std::size_t row_idx, std::size_t col_idx) const {
+    return data[row_idx * num_cols + col_idx];
+  }
+  __host__ __device__ std::size_t NumRows() const { return num_rows; }
+  __host__ __device__ std::size_t NumCols() const { return num_cols; }
+};
+
+
+}  // anonymous namespace
+
+namespace ML {
+namespace Explainer {
+
+template <typename ThresholdType, typename LeafType>
+ExtractedPathHandle
+extract_paths_impl(const tl::ModelImpl<ThresholdType, LeafType>& model) {
```

Comment:
The logic of this function seems generally correct, but it's also a bit hard to follow. Could we break it up a bit or consider using algorithms directly on the underlying containers where feasible rather than trying to keep track of all the separate indexes?